src/promql/src/extension_plan/normalize.rs

```diff
@@ -0,0 +1,355 @@
+            input,
+        }
+    }
+
+    pub fn to_execution_plan(&self, exec_input: Arc<dyn ExecutionPlan>) -> Arc<dyn ExecutionPlan> {
+        Arc::new(SeriesNormalizeExec {
+            offset: self.offset,
+            time_index_column_name: self.time_index_column_name.clone(),
+            input: exec_input,
+            metric: ExecutionPlanMetricsSet::new(),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct SeriesNormalizeExec {
+    offset: Millisecond,
+    time_index_column_name: String,
+
+    input: Arc<dyn ExecutionPlan>,
+    metric: ExecutionPlanMetricsSet,
+}
+
+impl ExecutionPlan for SeriesNormalizeExec {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.input.schema()
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        self.input.output_partitioning()
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        self.input.output_ordering()
+    }
+
+    fn maintains_input_order(&self) -> bool {
+        false
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.input.clone()]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
+        assert!(!children.is_empty());
+        Ok(Arc::new(Self {
+            offset: self.offset,
+            time_index_column_name: self.time_index_column_name.clone(),
+            input: children[0].clone(),
+            metric: self.metric.clone(),
+        }))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> DataFusionResult<SendableRecordBatchStream> {
+        let baseline_metric = BaselineMetrics::new(&self.metric, partition);
+
+        let input = self.input.execute(partition, context)?;
+        Ok(Box::pin(SeriesNormalizeStream {
+            offset: self.offset,
+            time_index_column_name: self.time_index_column_name.clone(),
+            schema: input.schema(),
+            input,
+            metric: baseline_metric,
+        }))
+    }
+
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default => {
+                write!(f, "PromSeriesNormalizeExec: offset=[{}]", self.offset)
+            }
+        }
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metric.clone_inner())
+    }
+
+    fn statistics(&self) -> Statistics {
+        self.input.statistics()
+    }
+}
+
+pub struct SeriesNormalizeStream {
+    offset: Millisecond,
+    time_index_column_name: String,
+
+    schema: SchemaRef,
+    input: SendableRecordBatchStream,
+    metric: BaselineMetrics,
+}
+
+impl SeriesNormalizeStream {
+    pub fn normalize(&self, input: RecordBatch) -> ArrowResult<RecordBatch> {
+        let ts_column_idx = self
```

Comment:
Looks like you could just hold the index instead of the name in `SeriesNormalizeStream`. `column_with_name()` is O(n), though isn't a performance issue at the time, it is a bit unnecessary as `self.schema` and `self.time_index_column_name` is available when we create the stream.